### ISSUE TITLE
Buidling challenge on Raspbian (latest/up2date)

### ISSUE THREAD

**Issue by aerno** [domoticz/domoticz#4296]
Hello,

I'm sorry to say that buidling has become a major challenge since stable 2020.2 on Raspbian. 

As the current release version, 2020.1, has major issues with the graphs of the energy measurement in the house (the graphs lose data, while the MQTT measurement has been up all the time)

Since the infamous 'extern' directory has been made fully remote since 2020.2, for me compiling a release version of Domoticz from scratch has become a drama.

First of all, it doesn't automatically download the libraries needed inside the extern directory:

----------------------------------------------------------------------------------

CMake Error at CMakeLists.txt:126 (message):
  The submodules were not downloaded! GIT_SUBMODULE was turned off or failed.
  Please update submodules and try again.

-- Configuring incomplete, errors occurred

----------------------------------------------------------------------------------

Okay I download the latest release version (or snapshot if no release available) of the 4 libraries needed manually and rename the output directories. No problem.

Then the installation part gives me a new error:
----------------------------------------------------------------------------------

CMake Error at extern/mosquitto/man/cmake_install.cmake:41 (file):
  file INSTALL cannot find
  "/abcfg/sources/domoticz/src/domoticz-2020.2/extern/mosquitto/man/mosquitto_passwd.1":
  No such file or directory.
Call Stack (most recent call first):
  extern/mosquitto/cmake_install.cmake:62 (include)
  cmake_install.cmake:47 (include)

----------------------------------------------------------------------------------

I will be able to make a patch which solves these installation errors and just skip these manual pages, but what good is a release version which needs so much tweaking to get the build to succeed ?

I will need to revert back to either 2020.1, where the graph content of the energy measurement is dropped randomly, or go back to the 2017 version of Domoticz (3.8153), which I've used for 2 years and is actually quite stable and has no graph data loss issues. Maybe it will last me another 2 years ;-)

As a last resort I will try to build a random  beta development snapshot by 'git clone' as see if that succeeds, but as people use Domoticz in production I wouldn't consider using a snapshot a good way to go.

I've been using Domoticz for about 4 years now and adivsed many friends and customers to use it, I wanted to inform your team that for me, as a Linux specialist, the building has become challenging and as usual there is no manual on how to build it (only for Windows) or for a beta snapshot online. 

Arno

Edit: Removed Issue Guidlines

**Comment by gizmocuz** [comment]
@aerno , You don't have to build Domoticz as binary versions are available and are up2date

As I flash on multiple raspbians and have no issue, try performing these steps:

Flash the latest Raspbian (Server) image on your raspberry pi
I suggest to use a Raspberry Pi 4 with 4-6GB memory, or a 3 at the lowest, 3 and lower needs swap memory

Perform the exact steps here:
https://www.domoticz.com/wiki/Installing_and_running_Domoticz_on_a_Raspberry_PI

**Comment by rwaaren** [comment]
@aerno 
I compile domoticz locally on a Raspberry 4 (buster) and on Intel NUC (1 buster / 1 bullseye) and cannot replicate any of the problems you describe. What (OS )version are you on ?

**Comment by gizmocuz** [comment]
@rwaaren , as mentioned, just use the latest raspbian image, start clean, and follow the wiki

**Comment by aerno** [comment]
I am on Raspbian Buster (up2date):

/home/arno> cat /etc/os-release 
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

**Comment by gizmocuz** [comment]
@aerno , I suspect you might have a free SD card to perform the above steps first

**Comment by aerno** [comment]
@gizmocuz  and @rwaaren  As my RPi runs more than just Domoticz (energy monitoring and control, mainly Mosquitto and my own Python apps) I cannot reinstall for every test. 

The hard thing with testing these features is that people have only one central energy meter in their house, one RFXcom and one Razberry. 

My system is an RPI3b+ having a 240GB SSD and 4GB of swap space, so for now the RPi is fine I think. 

Latest beta version seems to compile fine. Its release 2020.2 which isn't complete and ready to build.

**Comment by gizmocuz** [comment]
@aerno , that could indeed be the issue. There have been some recent changes in the CMake project.
Best to compile the beta for now, or use the CMake project file from the beta to build the stable.
It is also a good idea to rename the beta to stable soon.

**Comment by aerno** [comment]
I cannot access the SSD easily as I've switched to a USB based SSD solution, which has been built into a cabinet close the central energy distribution cabinet.

SDcards are notoriously unreliable on a RPi I think. Even when it has a perment 5V/5A Meanwell PSU, instead of a basic Wall-wart USB power supply which many people use. 

I have 3 RPis running here at our company and running them solely on SDcard will give them about 1-2 months of lifetime, then the card dies one way or another.

Domoticz doesn't write too much data, but it seems the wear leveling on sdcard is quite poor, so still it dies. 

Ofcourse I've switched off the last access time logging in ext4.

**Comment by aerno** [comment]
I have cmake 3.16.8 installed in /opt:

/home/arno> cmake --version
cmake version 3.16.8

Which I need for building Domoticz as the standard from Raspbian isn't new enough.

**Comment by aerno** [comment]
How come Domoticz depends on such a new CMake ? 

Is it the 'extern git' solution which needs that ? 

I couldn't believe that it is advised by Domoticz team to remove the standard cmake, which comes with Raspbian, and install a custom version in /usr/bin.

We're not on Windows, on Linux we have a package manager system :+1:

**Comment by kiddigital** [comment]
> 
> I couldn't believe that it is advised by Domoticz team to remove the standard cmake, which comes with Raspbian, and install a custom version in /usr/bin.
> 
> 
> 
> We're not on Windows, on Linux we have a package manager system :+1: 
> 
> 
@aerno Although I largely agree with your statement (btw like your user image 😁) I noticed over the years that for some packages, the generic distro's take quite a long time before an available GA version of something actually makes it into the packagemanager (update) list. Although some even have a beta channel with more recent version, still the gap is often months or more.

That is fine and arguably necessary (testing, dependencies, etc.) for 'major' packages, but for the group of 'smaller/simpler/specific' programs/packages it often just takes to long before a small fix or improvement made by someone in the community triples down to the distro's package manager and actually becomes useable by 'normal' users.

CMake is one of them (personally for me MPD is another example and there are many more)

Imagine how it would look like for Domoticz with new changes in the beta every few days and multiple GA's a year.

Just my 2 cents...

**Comment by gizmocuz** [comment]
I have spent enormous time developing this and keeping the wiki up to date
If you do not want to follow will please use the binaries

**Comment by aerno** [comment]
I appreciate all the time and effort that has been put into this project, thats also why I liked to share my concerns about building the project on the most common platform (I presume), the RPi with Raspbian.

As the project is so amazing I advised many people to use it, hence my input to the team of Raspbian. But if there is no need for input just let it be. 

Just to let you know, the latest github snapshot that I downloaded yesterday (I would call it version 20200821) builds fine on my up2date Raspbian setup. No need to reinstall or what have we.

Currently I have no option to go for a 'stable' release of Domoticz any more.

The advise to uninstall the original CMake is not necessary and in my opinion not desirable. How will people ever update the self installed cmake inside /usr/bin ? What if you want to add packages having a dependancy on cmake ? 

 I have the original Raspbian cmake version and cmake 3.16.8 installed in /opt. So there is no need for removing the original Raspbian cmake. 

@kiddigital Yes Linux distro builders take a long time to update packages, but for stability thats good I would say. Stable and dependable software makes for a reliable system.
